### DISCUSSION TITLE
Allow multiple apps to be deployed to the same server

### DISCUSSION THREAD

**Discussion by jaydrogers** [serversideup/spin#137] [Ideas]
# 👉 Describe the problem
[comment]: <> (What are you trying to solve?)

Spin only supports one app stack to be deployed to a server.

# 👥 Problem evidence & reach
[comment]: <> (How many people have this problem?)
- https://discord.com/channels/910287105714954251/910299290230997003/1316075190026895452

# 🥰 Describe the "impact" on users?
[comment]: <> (How will this make people's lives better once it is solved?)

This could save some users money and complexity so they don't need to manage multiple servers.

# 🏆 How to solve this problem
[comment]: <> (Describe the feature that you are proposing and how it will solve the problem)

### Create a "Spin Proxy" variable
This could be something enabled in the `.spin.yml` where users could enable "multitenancy". This would have the Ansible playbook create a default proxy (like Traefik). Once that "global service" is running, then users can use external networks to join to the proxy: https://docs.docker.com/compose/how-tos/networking/#use-a-pre-existing-network

# 💯 How do we validate the problem is solved?
[comment]: <> (Explain use cases on how we can measure the success of this implementation)
- [ ] A user can deploy multiple apps to the same server

**Comment by luchfirearmland** [comment]
Been waiting for this!!!! 🎉 (And InertiaJS SSR as well 🤣)

**Reply by vesper8** [reply]
Been waiting for this for 6 months too! Still waiting!

**Comment by jaydrogers** [comment]
My goal today was to get a better post put together, but I had a discussion with @harrisonratcliffe today and figured I would share these very quick notes. There is a chance as some of you chime in, I will put this into a better form of documentation and/or video.

## Background & TL;DR
I plan to automate and simplify this process natively in Spin, but it will take time to do this correctly. Instead of waiting I wanted to share notes that **it is possible to run multiple apps on a single server with Docker Swarm (what Spin uses) today**.

If you feel like tinkering, follow this quick guide.

## Important concepts
1.  When run a deployment with Spin on a server, it creates a Docker network with Traefik by default
2. Traefik can proxy requests to multiple apps using `labels:` differentiate where to send traffic
3. We can deploy a 2nd app to the same server, as long as we join it to an existing network that has Traefik running on it

More about the concept is here: https://docs.docker.com/compose/how-tos/networking/#use-an-existing-network

## Deploy and app with Spin
Run a usual deployment. It will have Traefik.

##  Create a 2nd app
You can create another app. It will have Traefik in the `docker-compose.prod.yml` file. What you want to do is *delete* any reference of Traefik in that config. We want to use the Traefik instance from the original app.

> [!IMPORTANT]  
> If you try to deploy Traefik to the same server, it will fail because another service is already using `HTTP (80)` and `HTTPS (443)`

## Adjust networks on the new app
You'll see in the `docker-compose.prod.yml` file has it's network definition:

```yaml
services:
  php:
    image: ${SPIN_IMAGE_DOCKERFILE_PHP}
    networks:
      - web-public
```

**We don't want to use `web-public`.**

Instead, SSH into your server and run `sudo docker network ls`. You might have a network called `spin_web-public` (or something like that). That's the "external network" we want to use.

So in our 2nd app, we want to change that network to EXACTLY MATCH and ALSO DECLARE that `spin_web-public` is an external network:

```yaml
services:
  php:
    image: ${SPIN_IMAGE_DOCKERFILE_PHP}
    networks:
      - spin_web-public
networks:
  spin_web-public:
    external: true

```

> [!IMPORTANT]  
> Make sure you update both the `network:` key and add the `networks:` key as I have above.

# Change the docker labels for Traefik
The next part is to get the Traefik configurations set, which is all done through `labels:`.

The original labels look like this:

```yml
      labels:
        - "traefik.enable=true"
        - "traefik.http.routers.laravel-web.rule=Host(`${SPIN_APP_DOMAIN}`)"
        - "traefik.http.routers.laravel-web.entrypoints=websecure"
        - "traefik.http.routers.laravel-web.tls=true"
        - "traefik.http.routers.laravel-web.tls.certresolver=letsencryptresolver"
        - "traefik.http.services.laravel-web.loadbalancer.server.port=8443"
        - "traefik.http.services.laravel-web.loadbalancer.server.scheme=https"
        # Health check
        - "traefik.http.services.laravel-web.loadbalancer.healthcheck.path=/up"
        - "traefik.http.services.laravel-web.loadbalancer.healthcheck.interval=30s"
        - "traefik.http.services.laravel-web.loadbalancer.healthcheck.timeout=5s"
        - "traefik.http.services.laravel-web.loadbalancer.healthcheck.scheme=https"
```

Notice a few key things here:

- `traefik.http.routers.laravel-web.rule` sets the domain, which is set by `${SPIN_APP_DOMAIN}`. You can try using `SPIN_APP_DOMAIN`, just make sure that variable is being set correctly (`example.com`).
- The Traefik labels need to be unique on a per app basis (`laravel-web` from above). If you deploy another app with the same labels, you'll have problems.

This means on your second app, you need to rename `laravel-web` to `laravel-web2` (or at least something different).

Again, make sure you're setting `SPIN_APP_DOMAIN` correctly in the process.

# Run a deployment
Deploy that sh*t and hopefully it will work 🤠
